<a href="https://colab.research.google.com/github/ramzymhafeel/SE4050-Deep-Learning-Lab-01/blob/main/notebooks/SE4050_Lab01_Task1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from IPython.display import display, Markdown

display(Markdown("""
# SE4050 - Deep Learning

## Lab 01: Introduction to Python Libraries for Deep Learning

### Task 1

**Student Name:** Mohamed Ramzy
**Index Number:** IT23386822
**Module:** SE4050 - Deep Learning
**Academic Year:** 2026
"""))


# SE4050 - Deep Learning

## Lab 01: Introduction to Python Libraries for Deep Learning

### Task 1

**Student Name:** Mohamed Ramzy  
**Index Number:** IT23386822  
**Module:** SE4050 - Deep Learning  
**Academic Year:** 2026


# SE4050 — Deep Learning

## Lab 01: Introduction to Python Libraries for Deep Learning

### Task 1

**Student Name:** Mohamed Ramzy
**Index Number:** IT23386822  
**Module:** SE4050 — Deep Learning  
**Academic Year:** 2026

## Objective

The objective of Task 1 is to apply NumPy, Matplotlib, Pandas, and Seaborn to:

- Generate random numerical data
- Visualize probability distributions
- Create a three-dimensional mathematical surface
- Calculate Pearson and Spearman correlations
- Visualize correlation matrices using heatmaps

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

print("Required libraries imported successfully.")

Required libraries imported successfully.


In [3]:
# Set a random seed so that the generated random values
# can be reproduced when the notebook is run again.

np.random.seed(42)

print("Random seed set to 42.")

Random seed set to 42.


## Task 1.1 — Random 4×4 Array from an Exponential Distribution

### Introduction

The exponential distribution is a continuous probability distribution commonly used to model the time between independent events.

In this task, NumPy is used to generate a random 4×4 array from an exponential distribution.

### Exponential Distribution Parameters

NumPy generates exponentially distributed random values using:

`np.random.exponential(scale, size)`

Where:

- `scale` is the inverse of the rate parameter
- `size` defines the output dimensions
- `size=(4, 4)` creates four rows and four columns

For this task, the scale value is set to `1.0`.

In [4]:
import numpy as np

# Set a fixed random seed for reproducible results.
np.random.seed(42)

print("Random seed set successfully.")

Random seed set successfully.


In [5]:
# Generate a random 4×4 array from an exponential distribution.

exponential_array = np.random.exponential(
    scale=1.0,
    size=(4, 4)
)

print("Random 4×4 array from an exponential distribution:")
print(exponential_array)

Random 4×4 array from an exponential distribution:
[[0.46926809 3.01012143 1.31674569 0.91294255]
 [0.16962487 0.16959629 0.05983877 2.01123086]
 [0.91908215 1.23125006 0.02079931 3.50355748]
 [1.78642954 0.23868763 0.20067899 0.20261142]]


In [6]:
# Display values rounded to four decimal places.

print("Exponential array rounded to four decimal places:\n")
print(np.round(exponential_array, 4))

Exponential array rounded to four decimal places:

[[0.4693 3.0101 1.3167 0.9129]
 [0.1696 0.1696 0.0598 2.0112]
 [0.9191 1.2313 0.0208 3.5036]
 [1.7864 0.2387 0.2007 0.2026]]


In [7]:
print("Array shape:", exponential_array.shape)
print("Number of dimensions:", exponential_array.ndim)
print("Total number of elements:", exponential_array.size)

Array shape: (4, 4)
Number of dimensions: 2
Total number of elements: 16


In [8]:
print("Array data type:", exponential_array.dtype)

Array data type: float64


In [9]:
array_mean = np.mean(exponential_array)
array_std = np.std(exponential_array)
array_min = np.min(exponential_array)
array_max = np.max(exponential_array)

print("Basic statistics of the generated array:")
print("Mean:", round(array_mean, 4))
print("Standard deviation:", round(array_std, 4))
print("Minimum value:", round(array_min, 4))
print("Maximum value:", round(array_max, 4))

Basic statistics of the generated array:
Mean: 1.0139
Standard deviation: 1.045
Minimum value: 0.0208
Maximum value: 3.5036


In [10]:
all_non_negative = np.all(exponential_array >= 0)

print("Are all values non-negative?", all_non_negative)

Are all values non-negative? True


In [11]:
import pandas as pd

array_df = pd.DataFrame(
    exponential_array,
    columns=["Column 1", "Column 2", "Column 3", "Column 4"],
    index=["Row 1", "Row 2", "Row 3", "Row 4"]
)

array_df.round(4)

,Column 1,Column 2,Column 3,Column 4
Row 1,0.4693,3.0101,1.3167,0.9129
Row 2,0.1696,0.1696,0.0598,2.0112
Row 3,0.9191,1.2313,0.0208,3.5036
Row 4,1.7864,0.2387,0.2007,0.2026


### Interpretation

A 4×4 NumPy array containing 16 random values was generated successfully from an exponential distribution.

The array contains only non-negative floating-point values, which is consistent with the properties of an exponential distribution.

The shape `(4, 4)` confirms that the generated array contains four rows and four columns.

In [12]:
# Task 1.1: Generate a random 4×4 array
# from an exponential distribution.

import numpy as np

np.random.seed(42)

exponential_array = np.random.exponential(
    scale=1.0,
    size=(4, 4)
)

print("Random 4×4 exponential-distribution array:\n")
print(np.round(exponential_array, 4))

print("\nArray shape:", exponential_array.shape)
print("Number of elements:", exponential_array.size)
print("All values are non-negative:",
      np.all(exponential_array >= 0))

Random 4×4 exponential-distribution array:

[[0.4693 3.0101 1.3167 0.9129]
 [0.1696 0.1696 0.0598 2.0112]
 [0.9191 1.2313 0.0208 3.5036]
 [1.7864 0.2387 0.2007 0.2026]]

Array shape: (4, 4)
Number of elements: 16
All values are non-negative: True


## Task 1.2 — Visualization of Exponential, Uniform, and Normal Distributions

## Task 1.3 — Three-Dimensional Plot of Z = X² + Y²

## Task 1.4 — Pearson and Spearman Correlation Analysis